OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs

```diff
@@ -182,60 +182,93 @@ public override bool HandleMouseInput(MouseInput mi)
 			return HandleEvent(icon, mi.Button == MouseButton.Left, mi.Modifiers.HasModifier(Modifiers.Shift));
 		}
 
-		bool HandleEvent(ProductionIcon icon, bool isLeftClick, bool handleMultiple)
+		protected bool PickUpCompletedBuildingIcon(ProductionIcon icon, ProductionItem item)
 		{
 			var actor = World.Map.Rules.Actors[icon.Name];
-			var first = icon.Queued.FirstOrDefault();
 
-			if (isLeftClick)
+			if (item != null && item.Done && actor.Traits.Contains<BuildingInfo>())
 			{
-				// Pick up a completed building
-				if (first != null && first.Done && actor.Traits.Contains<BuildingInfo>())
-				{
-					Sound.Play(TabClick);
-					World.OrderGenerator = new PlaceBuildingOrderGenerator(CurrentQueue, icon.Name);
-				}
-				else if (first != null && first.Paused)
-				{
-					// Resume a paused item
-					Sound.Play(TabClick);
-					World.IssueOrder(Order.PauseProduction(CurrentQueue.Actor, icon.Name, false));
-				}
-				else if (CurrentQueue.BuildableItems().Any(a => a.Name == icon.Name))
-				{
-					// Queue a new item
-					Sound.Play(TabClick);
-					Sound.PlayNotification(World.Map.Rules, World.LocalPlayer, "Speech", CurrentQueue.Info.QueuedAudio, World.LocalPlayer.Country.Race);
-					World.IssueOrder(Order.StartProduction(CurrentQueue.Actor, icon.Name,
-						handleMultiple ? 5 : 1));
-				}
-				else
-					Sound.Play(DisabledTabClick);
+				Sound.Play(TabClick);
+				World.OrderGenerator = new PlaceBuildingOrderGenerator(CurrentQueue, icon.Name);
+				return true;
 			}
-			else
+
+			return false;
+		}
+
+		public bool PickUpCompletedBuilding()
+		{
+			var myIcons = icons.Values.ToArray();
```

Comment:
This forces the whole icons collection to be walked and copied to a new object, while using LINQ will only enumerate as far as it needs to find the completed building.